pennylane/devices/qutrit_mixed/sampling.py

```diff
@@ -0,0 +1,367 @@
+    powers_of_three = QUDIT_DIM ** qml.math.arange(num_wires)[::-1]
+    indices = qml.math.array(samples @ powers_of_three)
+    return mp.eigvals()[indices]
+
+
+def _process_counts_samples(processed_sample, mp_has_obs):
+    """Processes a set of samples and counts the results."""
+    observables, counts = math.unique(processed_sample, return_counts=True, axis=0)
+    if not mp_has_obs:
+        observables = ["".join(observable.astype("str")) for observable in observables]
+    return dict(zip(observables, counts))
+
+
+def _process_expval_samples(processed_sample):
+    """Processes a set of samples and returns the expectation value of an observable."""
+    eigvals, counts = math.unique(processed_sample, return_counts=True)
+    probs = counts / math.sum(counts)
+    return math.dot(probs, eigvals)
+
+
+def _process_variance_samples(processed_sample):
+    """Processes a set of samples and returns the variance of an observable."""
+    eigvals, counts = math.unique(processed_sample, return_counts=True)
+    probs = counts / math.sum(counts)
+    return math.dot(probs, (eigvals**2)) - math.dot(probs, eigvals) ** 2
+
+
+# pylint:disable = too-many-arguments
+def _measure_with_samples_diagonalizing_gates(
+    mp: SampleMeasurement,
+    state: np.ndarray,
+    shots: Shots,
+    is_state_batched: bool = False,
+    rng=None,
+    prng_key=None,
+) -> TensorLike:
+    """Returns the samples of the measurement process performed on the given state,
+    by rotating the state into the measurement basis using the diagonalizing gates
+    given by the measurement process.
+
+    Args:
+        mp (~.measurements.SampleMeasurement): The sample measurement to perform
+        state (np.ndarray[complex]): The state vector to sample from
+        shots (~.measurements.Shots): The number of samples to take
+        is_state_batched (bool): whether the state is batched or not
+        rng (Union[None, int, array_like[int], SeedSequence, BitGenerator, Generator]): A
+            seed-like parameter matching that of ``seed`` for ``numpy.random.default_rng``.
+            If no value is provided, a default RNG will be used.
+        prng_key (Optional[jax.random.PRNGKey]): An optional ``jax.random.PRNGKey``. This is
+            the key to the JAX pseudo random number generator. Only for simulation using JAX.
+
+    Returns:
+        TensorLike[Any]: Sample measurement results
+    """
+    # apply diagonalizing gates
+    state = _apply_diagonalizing_gates(mp, state, is_state_batched)
+
+    total_indices = get_num_wires(state, is_state_batched)
+    wires = qml.wires.Wires(range(total_indices))
+
+    def _process_single_shot(samples):
+        samples_processed = _process_samples(mp, samples, wires)
+        if isinstance(mp, SampleMP):
+            return math.squeeze(samples_processed)
+        if isinstance(mp, CountsMP):
+            process_func = functools.partial(_process_counts_samples, mp_has_obs=bool(mp.obs))
```

Comment:
Something about treating an operator as a boolean feels wrong 😅 . Purely optional change:
```suggestion
            process_func = functools.partial(_process_counts_samples, mp_has_obs=mp.obs is not None)
```

Reply:
Makes sense, will change